### DISCUSSION TITLE
Transmission performance on Raspberry Pi

### DISCUSSION THREAD

**Discussion by 1100101** [transmission/transmission#5259] [Q&A] [answered]
Hi,

I downloaded a torrent today, and was a little surprised about the performance.
The Transmission process ran at 100% CPU, and interacting with the daemon via transmission-remote was slow/laggy.

Some details:
* I have a Raspberry Pi 4 with 1Gb of memory. At the time I checked about 130Mb were in use.
Nothing else of relevance was running at the time.
* I compiled transmission myself using the latest commit on the `main` branch, clang is the compiler used.
* there was a single torrent loaded, no other paused or seeding or downloading torrents
* The torrent had a size of about 2.5GB
* I have a 250Mbit DSL line, so 20Mb/s should be possible. I limit Transmission to 10MB/s, however, to not throttle other devices
* The torrent had at the time about 50 seeds and 20 leechers. According to transmission-remote I was downloading from about 46 peers.

My question is:
Is this expected behavior?
Where is all the CPU time going?
I remember downloading much larger torrents with similar/higher speeds, where I didn’t feel any lag between the daemon and the remote.
I have a feeling it’s because of the high number of peers I’m connected to?

**Comment by 1100101** [comment]
I must correct myself, the details about the torrents were not entirely true:
The torrent was actually ~35Gb in size, but I unselected all but 1 file in the file list.
So, I downloaded only one file with a size of 2.95Gb.

**Comment by ile6695** [comment]
Sorry for not being able to give a definitive answer.
Most of the time when there are slowdowns on desktop pc's, it's because of disk operations. If you use SD card that that is classified as class 10, it's rated for 10MB/s, but class 6 would be only 6MB/s and that could explain your slowdowns. I don't think the issue is high number of peers, but that could be tested by lowering the number of peers. [RPC-spec](https://github.com/transmission/transmission/blob/main/docs/rpc-spec.md) has information about how to limit that using remote.

IIRC raspbian has some profiling tools such as [hotspot](https://github.com/KDAB/hotspot) that can show you where that CPU time is being used.

**Comment by ckerr** [comment]
@1100101 please do use perf + hotspot to see where Transmission 4.0.x is spending its cycles. I would love to know if there are places where it can be improved.

Some helpful links on using these tools together:

https://www.kdab.com/hotspot-gui-linux-perf-profiler/

https://github.com/KDAB/hotspot#hotspot---the-linux-perf-gui-for-performance-analysis

**Comment by 1100101** [comment, marked as answer]
I’m embarrassed to say, the answer is much simpler, and totally my fault.
I switched from using gcc to clang, and forgot to set `CMAKE_BUILD_TYPE`.
With the option not set, clang defaults to `-O0`, and that’s what caused the bad performance.

With release mode properly set, CPU is now at around 50% (with peaks at 70%).
Transmission easily uses the whole bandwidth that’s available, and also the communication between the daemon and the remote is fast, without any delay.

Sorry for posting the question here, I should’ve checked my build settings before asking.
